### ISSUE TITLE
[rush] package version gets downgraded when `rush update --full` with pnpm v8

### ISSUE THREAD

**Issue by jeremymeng** [microsoft/rushstack#4283]
## Summary

- with pnpm v7, when we run `rush update --full` (with `"strictPeerDependencies": true` in pnpm-config.json), the latest version of dependency that satisfies the semver range is installed and saved in the lock file.
- however, with pnpm v8, `rush update --full`, in additional to switching to use v6 of the lockfile, also downgraded package versions to the exact version in the semver

## Repro steps

This is demostrated in this branch of test repo: https://github.com/iclanton/rush-test-repo/compare/main...jeremymeng:rush-test-repo:pnpm7

- add `"jssha": "^3.1.0"` to project3 package.json
- bump rush version to latest 5.102.0
- Set `"strictPeerDependencies": false` in pnpm-config.json
- `rush update --full` with pnpm v7, note that v5 lock file is generated, and jssha 3.31.0 installed and stored in pnpm-lock.yamlhttps://github.com/iclanton/rush-test-repo/commit/d5071ed3b5ab2b729622e8d8bc92821cd4a17efe
- change pnpm version to 8.6.12 in rush.json and run `rush update --full` again. Note that v6 lock file is used, but `jssha` 3.1.0 is installed and saved in lock file. https://github.com/iclanton/rush-test-repo/commit/9704d85214bf255e93ddc06cf9289d53fc2ba2f4

I'd expect jssha 3.31.0 too. Not sure if this is a pnpm behavior change or rush one, or combination of them.

Background:
we have a process that 
- Set `"strictPeerDependencies": false` in pnpm-config.json
- Set `"implicitlyPreferredVersions": true` in common-versions.json
- Run `rush update --full` and parse the output to see what version upgrades are available then create or update github workitems for those.

we also have a process that regularly runs `rush update --full` to upgrade to latest possible versions. 

The new behavior is breaking the above processes, because the lower versions are installed, and latest versions are no longer installed by `rush update --full`

are there other ways to achieve what we want? 1) upgrade to latest possible version of the semver 2) report versions available for upgrade but cannot done by 1) e.g., new major versions available.

  **Expected result:** 

  **Actual result:** 

## Standard questions

Please answer these questions to help us investigate your issue more quickly:

| Question | Answer |
| -------- | -------- |
| `@microsoft/rush` globally installed version? |  |
| `rushVersion` from rush.json? |  |
| `useWorkspaces` from rush.json? |  |
| Operating system? |  |
| Would you consider contributing a PR? |  |
| Node.js version (`node -v`)? |  |

**Comment by iclanton** [comment]
This is probably an issue with pnpm itself. Can you report an issue to https://github.com/pnpm/pnpm?

Is it possible to update your repo to use the workspaces feature? You'd probably experience fewer of these edge cases with that turned on.

**Comment by xirzec** [comment]
> Is it possible to update your repo to use the workspaces feature? You'd probably experience fewer of these edge cases with that turned on.

I would love to do this, but what's been holding us back is our monorepo contains hundreds of packages, so we aren't always able to take breaking changes in a dependency package across all consumers in one PR. 

My understanding is when workspaces is enabled, you can't have a version 2.x of a package in your tree and have other packages depend on version 1.x from the registry. That is to say, you can only have a single version of any package that is present in the monorepo.

**Comment by jeremymeng** [comment]
> This is probably an issue with pnpm itself

Right, looks that it's intentional https://github.com/pnpm/pnpm/issues/6463.  There's some workaround mentioned there too.

**Comment by octogonz** [comment]
Isn't this tunable using the [resolution-mode](https://pnpm.io/npmrc#resolution-mode) setting in common/config/rush/.npmrc?  

@iclanton / @dmichon-msft / @elliot-nelson  In your opinion, is the new default a good choice for a large monorepo?

We should probably consider:

1. Exposing `resolution-mode` as an official setting in [pnpm-config.json](https://rushjs.io/pages/configs/pnpm-config_json/)
2. Rush could offer a different default for `resolution-mode`

**Comment by dmichon-msft** [comment]
This is by design, based on the philosophy that if you specify a dependency on `foo@^1.0.0`, that means that your code must successfully build and test against `foo@1.0.0`. This avoids the class of issue in which you specify your dependency as `foo@^x`, and then write code that actually depends on features introduced in `foo@1.1.0`, such that when a consumer of your package performs a semantically valid installation and provides your package with `foo@1.0.0`, the package does not function as advertised.

The most correct thing to do would be to have multiple installations occur in your CI pipelines, including one at the minimum version and one at the maximum version, and ensure that all of them build and test correctly.

However, due to issues with both accidental breaking changes (`node-fetch` had one that broke Rush in a patch a while back, for example) and supply chain attacks (see `event-stream@3.3.6`), it is usually safer to _not_ automatically ingest a new version of a dependency without validating it first, thus default to lowest but have an automated flow to bump minimum versions in a timely manner.

When consuming a package, the version I want to know is what versions of the dependencies the package was developed and tested against, not what versions it might theoretically be compatible with if the entire ecosystem reliably obeyed semantic versioning (which many notable packages, e.g. typescript, don't).

Your preferences may vary if your monorepo doesn't _publish_ npm packages, only consume them.

TL;DR: Of the resolution modes currently available in pnpm, yes, I think `lowest-direct` is the most sensible default for a large monorepo.

**Comment by jeremymeng** [comment]
There's a chance that pnpm would revert the behavior https://github.com/pnpm/pnpm/issues/6463#issuecomment-1687169204

> The most correct thing to do would be to have multiple installations occur in your CI pipelines, including one at the minimum version and one at the maximum version, and ensure that all of them build and test correctly.

we have min/max testing currently only for @azure scoped packages. We had discussions about expanding them to our direct dependencies.

> it is usually safer to _not_ automatically ingest a new version of a dependency without validating it first, thus default to lowest but have an automated flow to bump minimum versions in a timely manner.

yes, we have a weekly automation that does that.  we could probably use the `resolution-mode` in this automation to get max versions.

**Comment by iclanton** [comment]
@xirzec - You can pull copies of local packages from the feed. You just need to list the packages that you want to pull from feed as `decoupledLocalDependencies` for the consuming projects in `rush.json`. See examples in this repo: https://github.com/microsoft/rushstack/blob/main/rush.json#L965

**Comment by iclanton** [comment]
Closing because this doesn't seem to actually be an issue with Rush.

**Comment by xirzec** [comment]
Yeah if we have 200 projects that depend on a common lib that just versioned, it's obnoxious to have to set this property for each of those packages only to just remove it again when we bump the version in package.json

It would be nicer if we could list which packages were allowed to be external if the version in the tree didn't satisfy the range.

**Comment by octogonz** [comment]
PNPM has now reverted their default:

![image](https://github.com/microsoft/rushstack/assets/4673363/b3a726d8-2f81-4929-9865-5d89c1396c48)

This is not the first time PNPM has changed a default setting only to revert in a non-MAJOR SemVer bump, responding to community feedback. These disruptions are probably negligible for the vast majority of small repos, but they can cause significant trouble for large monorepos with complex lockfiles and workflows, if the maintainers simply "upgraded" the PNPM version number without awareness of these nuanced consequences.

Rush [overrides some PNPM defaults](https://github.com/microsoft/rushstack/blob/40c535b17f7cd928f051ac65c8ceb0e60e355784/libraries/rush-lib/src/logic/base/BaseInstallManager.ts#L549C20-L549C22) and [even hardwires some settings](https://lfx.rushstack.io/pages/concepts/strict_settings/#settings-hardwired-by-rush), in order to provide a PNPM configuration that is better suited for large repos, as well as trying to minimize these disruptive impacts of PNPM upgrades.  The most recent example was [ignore-compatibility-db](https://rushjs.io/pages/advanced/compatibility_db/) (where PNPM's default behavior would tamper with **package.json** version specifiers using a fixup database that is invisible to users and can mysteriously force version changes in your lockfile when trying to upgrade `pnpmVersion`).

Therefore @iclanton @dmichon-msft  I'd like to reopen #4283 and reevaluate my questions from above:

### 1. PNPM 8's new default of lowest-direct, is it the best choice for a large monorepo?

> TL;DR: Of the resolution modes currently available in pnpm, yes, I think `lowest-direct` is the most sensible default for a large monorepo.

I'm not so sure about that, for a couple reasons: This default was new in PNPM 8 and now reverted, reflecting that it has caused confusion for PNPM users, as well as Rush users in this thread. We should think carefully before jumping to adopt it as our best practice.

> based on the philosophy that if you specify a dependency on `foo@^1.0.0`, that means that your code must successfully build and test against `foo@1.0.0`. This avoids the class of issue in which you specify your dependency as `foo@^1.0.0`, and then write code that actually depends on features introduced in `foo@1.1.0`, such that when a consumer of your package performs a semantically valid installation and provides your package with `foo@1.0.0`, the package does not function as advertised.

This philosophy assumes that NPM package maintainers will bump their SemVer range before publishing each release. For example, suppose `my-package` depends on `foo@^1.0.0` and my `package-lock.json` is actually using `foo@1.3.6` whereas the latest version is `foo@1.4.8`. The above philosophy assumes that, prior to publishing `my-package`, I will bump my `my-package/package.json` to depend on `foo@^1.3.6` (not `foo@^1.4.8`) communicating that `1.3.6` is the latest version that I tested with.

Is that what people really do? No, I think the vast majority of package maintainers are lazy. They will never modify the `"^1.0.0"` string unless they are explicitly upgrading for some important reason.

In fact what we **_really_** want seems like [resolution-mode=time-based](https://pnpm.io/npmrc#resolution-mode):

| `resolution‑mode=` |version|meaning|
|---|---|---|
|`highest`|1.4.8|The highest version of `foo` satisfying `^1.0.0`|
|`lowest`|1.0.0|The lowest version satisfying `^1.0.0`|
|`time-based`|1.3.6| The latest version of `foo` satisfying `^1.0.0` at the time when `my-package` was published|

Why isn't `time-based` the default then? I don't know, but it's probably a complicated question with counterintuitive tradeoffs. So we should think carefully before changing the status quo.

### 2. Exposing `resolution-mode` as an official setting in [pnpm-config.json](https://rushjs.io/pages/configs/pnpm-config_json/)

👉 Given the confusion in this thread, I think at least this work item must be completed before we can close this issue #4283.

### 3. Should Rush try to protect our users from getting disrupted by the reverted policy decisions in `8.0.0`-`8.6.12`?

We could consider providing a stable default (while still honoring an explicit `resolution-mode` setting in **pnpm-config-.json**, if users chose to specify it).

Or we could just print a warning as we did for **pnpmCompatibilityDb**:

https://github.com/microsoft/rushstack/blob/40c535b17f7cd928f051ac65c8ceb0e60e355784/libraries/rush-lib/src/logic/base/BaseInstallManager.ts#L600-L612

**Comment by octogonz** [comment]
The PNPM docs provide some further details about `resolution-mode=time-based`:

> With this resolution mode installations with warm cache are faster. It also reduces the chance of subdependency hijacking as subdependencies will be updated only if direct dependencies are updated.
> 
> This resolution mode works only with npm's [full metadata]. So it is slower in some scenarios. However, if you use [Verdaccio] v5.15.1 or newer, you may set the `registry-supports-time-field` setting to `true`, and it will be really fast.
> 
> When `resolution-mode` is set to `lowest-direct`, direct dependencies will be resolved to their lowest versions.
>
> [full metadata]: https://github.com/npm/registry/blob/master/docs/responses/package-metadata.md#full-metadata-format
> [Verdaccio]: https://verdaccio.org/

"Slower in some scenarios" might be why it's not the default.
